### PR TITLE
MINOR: make sure all generated data tests cover all versions

### PR DESCRIPTION
**Pull request by chia7712**
inspired by #10068

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java [hunk 1]
```diff
@@ -201,7 +201,7 @@ public short oldestVersion() {
 
     public List<Short> allVersions() {
         List<Short> versions = new ArrayList<>(latestVersion() - oldestVersion() + 1);
-        for (short version = oldestVersion(); version < latestVersion(); version++) {
+        for (short version = oldestVersion(); version <= latestVersion(); version++) {
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java:204]: This is not a critical bug as the method `allVersions` is used by testing only.

File: clients/src/test/java/org/apache/kafka/common/requests/OffsetsForLeaderEpochRequestTest.java [hunk 2]
```diff
@@ -34,15 +34,15 @@ public void testForConsumerRequiresVersion3() {
             assertThrows(UnsupportedVersionException.class, () -> builder.build(v));
         }
 
-        for (short version = 3; version < ApiKeys.OFFSET_FOR_LEADER_EPOCH.latestVersion(); version++) {
+        for (short version = 3; version <= ApiKeys.OFFSET_FOR_LEADER_EPOCH.latestVersion(); version++) {
             OffsetsForLeaderEpochRequest request = builder.build((short) 3);
```
Comment (dajac) [inline clients/src/test/java/org/apache/kafka/common/requests/OffsetsForLeaderEpochRequestTest.java]: Should we replace `3` by `version` here?

File: clients/src/test/java/org/apache/kafka/common/requests/ProduceRequestTest.java [hunk 3]
```diff
@@ -291,19 +292,11 @@ public void testMixedIdempotentData() {
...
-    }
-
-    private void assertThrowsInvalidRecordException(ProduceRequest.Builder builder, short version) {
-        try {
-            builder.build(version).serialize();
-            fail("Builder did not raise " + InvalidRecordException.class.getName() + " as expected");
-        } catch (RuntimeException e) {
-            assertTrue(InvalidRecordException.class.isAssignableFrom(e.getClass()),
-                "Unexpected exception type " + e.getClass().getName());
+    private static <T extends Throwable> void assertThrowsInvalidRecordExceptionForAllVersions(ProduceRequest.Builder builder,
```
Comment (dajac) [inline clients/src/test/java/org/apache/kafka/common/requests/ProduceRequestTest.java]: Should we rename the method now that the exception is passed as an argument?

File: clients/src/test/java/org/apache/kafka/common/requests/ProduceRequestTest.java [hunk 4]
```diff
@@ -291,19 +292,11 @@ public void testMixedIdempotentData() {
...
+                                                                                               Class<T> expectedType) {
+        for (short version = builder.oldestAllowedVersion(); version <= builder.latestAllowedVersion(); version++) {
+            short v = version;
```
Comment (dajac) [inline clients/src/test/java/org/apache/kafka/common/requests/ProduceRequestTest.java]: Could we get rid of `v` and use `version` below directly?

File: clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java [hunk 5]
```diff
@@ -295,7 +295,7 @@ public void testSerialization() throws Exception {
         checkRequest(createDeleteGroupsRequest(), true);
         checkErrorResponse(createDeleteGroupsRequest(), unknownServerException, true);
         checkResponse(createDeleteGroupsResponse(), 0, true);
-        for (int i = 0; i < ApiKeys.LIST_OFFSETS.latestVersion(); i++) {
+        for (int i = 0; i <= ApiKeys.LIST_OFFSETS.latestVersion(); i++) {
```
Comment (dajac) [inline clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java]: Could we use `allVersions` here as well?

File: clients/src/test/java/org/apache/kafka/common/message/MessageTest.java [hunk 6]
```diff
@@ -485,7 +485,7 @@ public void testOffsetCommitRequestVersions() throws Exception {
             if (version == 1) {
                 testEquivalentMessageRoundTrip(version, requestData);
             } else if (version >= 2 && version <= 4) {
-                testAllMessageRoundTripsBetweenVersions(version, (short) 4, requestData, requestData);
+                testAllMessageRoundTripsBetweenVersions(version, (short) 5, requestData, requestData);
```
Comment (dajac) [inline clients/src/test/java/org/apache/kafka/common/message/MessageTest.java:488]: Why are we changing this?

**Review by dajac — COMMENTED** [review]
Thanks for the PR. I have left few comments.

File: clients/src/test/java/org/apache/kafka/common/requests/OffsetsForLeaderEpochRequestTest.java (hunk 2 above)
Comment (chia7712) [reply clients/src/test/java/org/apache/kafka/common/requests/OffsetsForLeaderEpochRequestTest.java]: good catch. fixed.

File: clients/src/test/java/org/apache/kafka/common/requests/ProduceRequestTest.java (hunk 4 above)
Comment (chia7712) [reply clients/src/test/java/org/apache/kafka/common/requests/ProduceRequestTest.java]: It requires final variable. I rewrite code by lambda foreach to eliminate duplicate intermediate variable.

File: clients/src/test/java/org/apache/kafka/common/message/MessageTest.java (hunk 6 above)
Comment (chia7712) [reply clients/src/test/java/org/apache/kafka/common/message/MessageTest.java:488]: `testAllMessageRoundTripsBetweenVersions` exclude the end number so it should pass 5 so as to test version 4

**Review by dajac — APPROVED** [review]
LGTM, thanks for the PR.

**Pull request merged**